### ISSUE TITLE
CNFE in ScriptModuleActivator

### ISSUE THREAD

**Issue by maggu2810** [eclipse/smarthome#3013]
I get a CNFE on the first Script rule execution.

```text
Exception while importing scope: java.lang.ClassNotFoundException: org.eclipse.smarthome.core.library.types.OpenClosedType cannot be found by com.eclipsesource.jaxrs.jersey-min_2.22.1
```

Details below...

* Use the recent master branch (commit f3f43bf3c764bb03e7465139c17a75cf3c058f34) without local modifications
* Change the Eclipse SmartHome launch configuration and enable the most of the Automation bundles
* Start the Eclipse SmartHome launch configuration
* Open the `Rules` section of the PaperUI
* Create a new rule
* Choose a name
* Add an action by pressing on `+` after the `then...` text
* Select the action type `execute a given script`
* Press on `NEXT`
* Check that the selectio is set to `JavaScript`
* Add a simple script e.g. `print("hello");`
* Press on `OK`
* Save the rule
* The log should contain something similar to

```text
18:40:19.528 INFO  smarthome.event.RuleAddedEvent[:43] - Rule 'rule_1' has been added.
18:40:19.550 INFO  s.event.RuleStatusInfoEvent[:43] - rule_1 updated: INITIALIZING
18:40:19.564 INFO  s.event.RuleStatusInfoEvent[:43] - rule_1 updated: IDLE
18:40:19.568 DEBUG o.e.s.c.c.r.AbstractManagedProvider[:62] - Added new element rule_1 to ManagedRuleProvider.
```

* Press the `Run Now` button

Now the Console shows me a `ClassNotFoundException`.
This exception is only shown the first time and regardless of that excetion the rule is executed.

```
18:41:03.966 INFO  s.event.RuleStatusInfoEvent[:43] - rule_1 updated: RUNNING
18:41:04.318 WARN  o.e.s.a.m.s.i.ScriptModuleActivator[:211] - Found wrong classloader: To prevent class loading problems use this directive: -Dorg.osgi.framework.bundle.parent=ext
18:41:04.318 DEBUG o.e.s.a.m.s.i.ScriptModuleActivator[:214] - initializing script scope from 'DefaultScriptScopeProvider' for engine 'Oracle Nashorn'.
18:41:04.481 ERROR o.e.s.a.m.s.i.ScriptModuleActivator[:231] - Exception while importing scope: java.lang.ClassNotFoundException: org.eclipse.smarthome.core.library.types.OpenClosedType cannot be found by com.eclipsesource.jaxrs.jersey-min_2.22.1
hello
18:41:04.487 DEBUG o.e.s.a.c.internal.RuleEngine[:882] - The rule 'rule_1' is executed.
18:41:04.488 INFO  s.event.RuleStatusInfoEvent[:43] - rule_1 updated: IDLE
```

**Comment by danchom** [comment]
The problem should be in script action handler. As it is seen in the logs the runNow command is trying to execute the action and there is an exception in javascript evaluation.

**Comment by maggu2810** [comment]
I don't think so.
I assume it is caused because the Script Engine implementation (in this case Nashorn) tries to load classes that are not part of its class loader.
Providers inject classes that are not part of the class loader used by the script module activator (not declared in its manifest etc.) so how to you think the respective script engine implementation should know about?
Does the script engine use the thread context classloader?
If it is using that classloader we could create a unified classloader that contains all classloader of the used providers...

**Comment by maggu2810** [comment]
But perhaps I should do what the comment state:
```
Found wrong classloader: To prevent class loading problems use this directive: -Dorg.osgi.framework.bundle.parent=ext
```

Will give it a try 😉

**Comment by maggu2810** [comment]
As written here http://stackoverflow.com/a/30251930/7546824 is will work using a modified context class loader in engine create using the factory method.

To test it I modified the `getScriptEngine` method of `ScriptModuleActivator`.

```java
    public static synchronized ScriptEngine getScriptEngine(String type) {
        ScriptEngine engine = engines.get(type);
        if (engine == null) {
            final Collection<ClassLoader> classloaders = new HashSet<>();
            for (ScriptScopeProvider provider : scriptScopeProviders) {
                classloaders.add(provider.getClass().getClassLoader());
                for (final Object obj : provider.getScopeElements().values()) {
                    classloaders.add(obj.getClass().getClassLoader());
                }
            }
            final Thread currentThread = Thread.currentThread();
            final ClassLoader ccl = currentThread.getContextClassLoader();
            JoinClassLoader jcl = new JoinClassLoader(ccl, classloaders);
            currentThread.setContextClassLoader(jcl);
            try {
                engine = engineManager.getEngineByMimeType(type);
            } catch (Exception e) {
                logger.error("Exception while get engine by mime type: {}", e.getMessage());
            } finally {
                currentThread.setContextClassLoader(ccl);
            }
            for (ScriptScopeProvider provider : scriptScopeProviders) {
                initializeScope(engine, provider);
            }
            engines.put(type, engine);
        }
        return engine;
    }
```

No class loading issue anymore.

**Comment by kaikreuzer** [comment]
Note that this part of the code is refactored in https://github.com/eclipse/smarthome/pull/1783 anyhow - you should check if it works in there as expected.

**Comment by maggu2810** [comment]
If we don't fix this because it is refactored by #1783 I don't think we should postpone that PR because of this issue.
As it is not solved correctly in the current code base, so #1783 does not break it, we can fix it after #1783 has been merged.
I don't see any class loader modification / injection in the referenced PR, so I assume it is not solved there, too.

**Comment by kaikreuzer** [comment]
As `ScriptModuleActivator` does not exist anymore meanwhile, I would like to close this issue - if there is still a problem, it will likely be a different one :-)
